### PR TITLE
Perimeter insulation for slabs on grade (NECB)

### PR DESCRIPTION
**Pull request by brgix**
_Linked to BTAP Task [497](https://github.com/canmet-energy/btap_tasks/issues/497) (for those who have access). Work in progress. Should be completed/tested/reviewed/merged within a few days._

----
**Background**

NECB editions (2011 through 2025) hold the very same insulation requirements for surfaces in contact with the **ground**:

Image

For **slabs on grade** (i.e. "floors" in the above table), insulation requirements are limited to 4ft (panels) along the exterior perimeter, except for climate zone 8 (where continuous insulation is instead required).

Image

BTAP systematically adds continuous insulation for all slabs-on-grade (regardless of climate zone), despite relying on EnergyPlus' KIVA option for ground heat loss (KIVA supports perimeter insulation). Enhancing BTAP's KIVA definitions is fairly straightforward. Estimating the resulting perimeter insulation surface area (for costing, embodied carbon) is a tad more complicated, requiring geometry operations that are similar to those estimating a space's [width](https://github.com/NatLabRockies/openstudio-standards/blob/cd105f8e8f5d0980b5d0bb2468ec89c9aa517f04/lib/openstudio-standards/btap/geometry.rb#L535).

Image

### CODE REVIEW

File: lib/openstudio-standards/standards/necb/common/construction_defaults.osm [hunk 1]
```diff
@@ -1344,11 +1344,14 @@ OS:DefaultConstructionSet,
   ,                                       !- Building Shading Construction Name
   ;                                       !- Site Shading Construction Name
 
-OS:Material:NoMass,
+OS:Material,
   {8123ea19-659f-408b-9bbf-9d3d1a8354b5}, !- Handle
   Typical Carpet Pad,                     !- Name
-  Smooth,                                 !- Roughness
-  0.216479986995276,                      !- Thermal Resistance {m2-K/W}
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/common/construction_defaults.osm:1351]: There are 2 instances of _massless_ (or "NoMass") materials in this BTAP/NECB (common resource) OSM file. BTAP methods (like [replace_massless_material_with_std_material](https://github.com/NatLabRockies/openstudio-standards/blob/dbf99c96c6ec98a6bf43a5c090236a551312540d/lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb#L1007)) were subsequently developed to convert such materials into standard materials. Here, the massless materials are instead replaced by standard ones at the get go. No need for any subsequently fix.

File: lib/openstudio-standards/standards/necb/NECB2011/building_envelope.rb [hunk 2]
```diff
@@ -1189,10 +1189,13 @@ def model_add_construction_set_from_osm(model:,
   end
 
   def assign_contruction_to_adiabatic_surfaces(model)
-    cp02_carpet_pad = OpenStudio::Model::MasslessOpaqueMaterial.new(model)
+    cp02_carpet_pad = OpenStudio::Model::StandardOpaqueMaterial.new(model)
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/building_envelope.rb:1192]: Avoiding the use of _massless_ materials altogether in BTAP. Both NECB unit tests are green (for a wide range of cases):

- test_necb_structures.rb
- test_NECB_TBD.rb

File: lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb [hunk 3]
```diff
@@ -575,7 +575,7 @@ def model_apply_standard(model:,
                       sizing_run_dir: sizing_run_dir,
                       lights_type: lights_type,
                       lights_scale: lights_scale)
-    apply_kiva_foundation(model, massive)
+    apply_kiva_foundation(model)
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb:578]: No longer any need to distinguish between _current_ vs _updated_ BTAP approaches to construction/material generation.

File: lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb [hunk 4]
```diff
@@ -1105,81 +1105,79 @@ def set_construction_air_film_resistances(model: nil)
...
-          else
-            replace_massless_material_with_std_material(model,gfloor)
-            gfloor.setOutsideBoundaryCondition('Foundation')
-            gfloor.setAdjacentFoundation(zone_kiva_models.last)
-          end
+
+          c = gfloor.construction.get.to_LayeredConstruction.get
+          gfloor.setOutsideBoundaryCondition('Foundation')
+          gfloor.setAdjacentFoundation(zone_kiva_models.last)
+          gfloor.setConstruction(c)
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb:1205]: Reiterating that there is no longer any need to distinguish between _current_ vs _updated_ BTAP approaches to construction/material generation - no longer any _massless_ materials to deal with.

File: lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb [hunk 5]
```diff
@@ -1105,81 +1105,79 @@ def set_construction_air_film_resistances(model: nil)
...
+              gwall.setAdjacentFoundation(zone_kiva_models.last)
+              gwall.setConstruction(c)
             end
           end
         end
       end
     end
-    kiva_settings = model.getFoundationKivaSettings if !model.getFoundationKivas.empty?
+
+    model.getFoundationKivaSettings
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb]: This may have changed in recent SDK releases (?), but every OpenStudio model has a valid _OpenStudio::Model::FoundationKivaSettings_ instance. Returning this variable for the moment - may soon change.

File: lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb [hunk 6]
```diff
@@ -1131,6 +1133,29 @@ def apply_kiva_foundation(model)
...
+              xps38.setThickness(0.038)
+              xps38.setConductivity(0.029)
+              xps38.setDensity(28)
+              xps38.setSpecificHeat(1450)
+            else
+              xps38 = xps38.get
+            end
+
+            zone_kiva_models.last.setInteriorHorizontalInsulationMaterial(xps38)
+            zone_kiva_models.last.setInteriorHorizontalInsulationWidth(1.2)
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/necb_2011.rb:1157]: Solution adds 1-1/2" thick XPS insulation along slab perimeter (4ft in width) if:

- at least one outdoor-facing wall along slab perimeter
- HDD < 7000

_Structure_ & _Bridging_ units remain green. Spot checks of generated _in.osm_ and _run/eplusout.err_ files are void of any errors.

File: lib/openstudio-standards/btap/geometry.rb [hunk 7]
```diff
@@ -165,6 +165,54 @@ def self.space_width(space = nil)
...
+        offset = OpenStudio.buffer(polyg, -w, 0.01)
+        return 0 if offset.empty?
+
+        m2 = OpenStudio.getArea(offset.get)
+
+        puts area
+        puts m2
+        puts
+
+        m2.empty? ? 0 : area - m2.get
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb]: **EDIT**: _Solution now works for these cases, and most others._

----

For around half of the prototype models, the current calculations yield correct results. For the moment, the solution fails for many models, like the _Secondary School_:

Image

The solution will also need to handle multiple pavilions, e.g. _LEEP Town House_:

Image

... but here, this will likely boil down to handling multiple, distinct _unions_ of ground-facing floors.

File: lib/openstudio-standards/btap/geometry.rb [hunk 8]
```diff
@@ -210,6 +210,37 @@ def self.perimeter_m2(spaces = [], w = 1.2)
         puts area
         puts m2
         puts
+        # Prototype                 slab  union?     cutout   perimeter OK?
+        # FullServiceRestaurant   511.15     OK     408.39      102.76  OK
+        # HighriseApartment       783.65     OK     637.63      146.02
+        # HighriseApartmentMult   783.65     OK     637.63      146.02
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb]: For each BTAP Prototype model:

- **slab** (m2) : (expected) area of (merged) ground-facing slabs, using **SketchUp** _offset_ 
- **union?** : whether area of merged slabs (using **OpenStudio.joinAll**) matches "slab (m2)"
- **cutout** (m2) : (expected) area of _offset_ polygon (by 1.2m), as reported by **SketchUp**
- **perimeter** (m2): (expected) area of _perimeter_ (slab - cutout), as reported by **SketchUp**
- **OK?** : whether the area of the **OpenStudio.buffer** generated _cutout_ matches **SketchUp**

Setting aside the _NorthernEducation_ and _NorthernHealthCare_ models (no ground facing slabs), the LargeOffice (full basement), as well as the _LEEPTownHouse_ model (2 separate pavilions, see previous comment).

File: lib/openstudio-standards/btap/geometry.rb [hunk 9]
```diff
@@ -167,81 +167,114 @@ def self.space_width(space = nil)
...
       # @return [Numeric] perimeter area (0 if failure)
       def self.perimeter_m2(spaces = [], w = 1.2)
-        slabs  = []
-        slabs += TBD.facets(spaces, "foundation", "floor")
-        slabs += TBD.facets(spaces, "ground", "floor")
+        slabs = []
+
+        # Fetch slabs-on-grade.
+        spaces.each do |space|
+          t = space.siteTransformation
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:201]: _For Pete's sake!_ Many models have spaces with relative XYZ origins. Factoring _site transformation_ fixes most cases.

File: lib/openstudio-standards/btap/geometry.rb [hunk 10]
```diff
@@ -167,81 +167,114 @@ def self.space_width(space = nil)
...
 
-          if slabs.size > 1
-            polygs = slabs.map     { |slb| TBD.poly(slb, false, true, true, t, :ulc) }
-            polygs = polygs.reject { |plg| plg.empty? }
-            polygs = polygs.map    { |plg| plg.to_a.reverse }
+        # OpenStudio's 'joinAll' is sensitive to the order/sequence of polygons.
+        # A model may hold 20 slabs: a first call to 'joinAll' may successfully
+        # merge all 20 into 3 groups. A follow-up call may finally unite all.
+        while slabs.size != 1
+          break if ctr > 100
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:231]: Attempts to 'join' as little as 4 or 6 slabs into a single polygon can produce incomplete merges. A few additional iterations may be required.

File: lib/openstudio-standards/btap/geometry.rb [hunk 11]
```diff
@@ -167,81 +167,114 @@ def self.space_width(space = nil)
...
+          slabs = OpenStudio.joinAll(slabs.shuffle, 0.01)
+          ctr  += 1
         end
 
-        offset = OpenStudio.buffer(polyg, -w, 0.01)
-        return 0 if offset.empty?
+        total = 0
+
+        # If multiple unions, possibly dealing with separate pavillions. Add up.
+        slabs.each do |slab|
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:238]: The _LEEPTownHouse_ is modelled as 2 pavilions. Looping through each group (of slabs) produces the correct result.

File: lib/openstudio-standards/btap/geometry.rb [hunk 12]
```diff
@@ -167,81 +167,114 @@ def self.space_width(space = nil)
...
+        # SecondarySchool       11902.00     OK   11012.56      889.44  OK
         # SmallHotel             1003.40     OK     833.59      169.81  OK
         # SmallOffice             511.16     OK     406.16      105.00  OK
         # Warehouse              4598.25     OK    4252.90      345.35  OK
         #
-        # SPECIAL CASES:
+        # LEEPTownHouse1          466.15     OK     364.44      101.71  OK (Pavillion 1)
+        # LEEPTownHouse2          699.22     OK     563.52      135.70  OK (Pavillion 2)
+        # ---------------------- ------- ------- ---------- ---------------
+        # LEEPTownHouse TOTAL    1165.37     OK     927.96      237.41  OK
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb]: Most prototypes are now handled correctly, including the multiple pavilions of the _LEEPTownHouse_.

File: lib/openstudio-standards/btap/geometry.rb [hunk 13]
```diff
@@ -167,81 +167,114 @@ def self.space_width(space = nil)
...
-        # LEEPTownHouse1          466.15            364.44      101.71  ... (1st pavillion)
-        # LEEPTownHouse2          699.22            563.52      135.70  ... (2nd pavillion)
         #
-        # NortherEducation   (no ground-facing floors ... facing outdoors)
-        # NorthernHealthCare (no ground-facing floors ... facing outdoors)
+        # PROBLEM CASES (Boost feedback: "union has inner loops"):
+        # ----------------------------------------------------------------------
+        # Prototype          slab         union?          cutout
+        # LEEPMultiTower  2814.72 (2441)  2434.49 (2043)  380.23 (397) 
+        # Outpatient      1373.29 (1199)  1164.31 ( 875)  208.98 (324)
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb]: Two problem cases. _Boost_ complains of inner loops, which sounds like invalid geometry to me - haven't checked yet (TO-DO).

BTAP is supposed to handle 3rd-party models, without NRCan supervision. As there is little to no validation of geometry in BTAP (pre-OpenStudio/EnergyPlus runs), the solution may be to _wing it_. The (incorrect) results (for costing) are not that off the mark ...

File: test/necb/unit_tests/tests/test_necb_dimensions.rb [hunk 14]
```diff
@@ -100,4 +98,118 @@ def test_necb_dimensions()
...
+    # LargeOffice        : full basement
+    # Hospital           : full basement
+    # NortherEducation   : no ground-facing floors
+    # NorthernHealthCare : no ground-facing floors
+    #
+    #
+    # PROBLEM CASES (doesn't work initially, works after BTAP modifications):
+    # ----------------------------------------------------------------------
+    # LargeHotel
+    # LEEPTownHouse
```
Comment (brgix) [inline test/necb/unit_tests/tests/test_necb_dimensions.rb]: Both prototypes yield correct results post-BTAP (e.g. running unit `test_necb_structures.rb`), yet fail when running `BTAP::Geometry::Spaces.perimeter_m2` on the original, unaltered models. Will need to take a deeper dive ...

File: lib/openstudio-standards/btap/geometry.rb [hunk 15]
```diff
@@ -178,8 +178,24 @@ def self.space_width(space = nil)
...
 
+        spaces = spaces.is_a?(OpenStudio::Model::Space) ? [spaces] : spaces
+        spaces = spaces.respond_to?(:to_a) ? spaces.to_a : []
+        return [] if spaces.empty?
+
+        bldg = spaces.first.model.getBuilding
+
+        # Check if pre-set as AdditionalProperty.
+        if bldg.additionalProperties.hasFeature(tag)
+          aire = bldg.additionalProperties.getFeatureAsDouble(tag)
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:192]: Likely pre-setting perimeter insulation area as AdditionalProperty. This would be caught here (and returned). The method remains useful for 3rd-party models running BTAP.

File: lib/openstudio-standards/btap/geometry.rb [hunk 16]
```diff
@@ -234,6 +249,13 @@ def self.perimeter_m2(spaces = [], w = 1.2)
           total += area.get - m2.get
         end
 
+        # Add as AdditionalProperty (if missing) for future calls.
+        if total > 0
+          unless bldg.additionalProperties.hasFeature(tag)
+            bldg.additionalProperties.setFeature(tag, total)
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:257]: Setting slab-on-grade perimeter area as AdditionalProperty (model _building_), if applicable.

File: test/necb/unit_tests/tests/test_necb_dimensions.rb [hunk 17]
```diff
@@ -172,44 +172,58 @@ def test_necb_slab_perimeters()
...
-
-      emsg = "BTAP/Dimensions: slab perimeter (#{cas})"
-      assert_in_delta(m2, area, 0.02, emsg)
+    fdback << "BTAP/Dimensions Slab Unit Tests"
+    fdback << "~~~~~~~~~~~~~~~ ~~~~ ~~~~ ~~~~~"
+
+    # Increasing the number of 'times' (e.g. to 100) increases the likely of
+    # catching failed slab surface merges. Failures seem limited to the paired
+    # pavilions of the 'LEEPTownHouse' prototype (sometimes getting 220.41 m2).
+    1.times do |i|
```
Comment (brgix) [inline test/necb/unit_tests/tests/test_necb_dimensions.rb:91]: Further testing shows the (dual pavilion) _LEEPTownHouse_ yields either 220.41 m2 (wrong answer) or 237.41 m2 (right answer). Can't pinpoint _why?_ This is the only model with this issue, however.

File: test/necb/unit_tests/tests/test_necb_dimensions.rb [hunk 18]
```diff
@@ -153,15 +153,15 @@ def test_necb_slab_perimeters()
     # SKIPPED CASES (as expected):
     # ----------------------------------------------------------------------
     # LargeOffice        : full basement
+    # LargeHotel         : full basement
```
Comment (brgix) [inline test/necb/unit_tests/tests/test_necb_dimensions.rb]: Yup. _LargeHotel_ has a full basement - my bad.

File: lib/openstudio-standards/btap/geometry.rb [hunk 19]
```diff
@@ -197,17 +197,19 @@ def self.perimeter_m2(spaces = [], w = 1.2)
...
 
         # Fetch slabs-on-grade.
-        spaces.sort.each do |space|
+        spaces.each do |space|
           t = space.siteTransformation
 
           wlls  = []
           wlls += TBD.facets(space, "ground", "wall")
           wlls += TBD.facets(space, "foundation", "wall")
+          wlls += TBD.facets(space, "othersidecoefficients", "wall")
```
Comment (brgix) [inline lib/openstudio-standards/btap/geometry.rb:206]: A few of the Prototype models have "OtherSideCoefficients" as outside boundary conditions for basement wall + slab entries. This is later reset (e.g. to "Foundation") by BTAP. Catching "OtherSideCoefficients" here (earlier in the process) to simplify unit tests (no need to run EnergyPlus sizing runs).

File: lib/openstudio-standards/standards/necb/NECB2011/data/geometry/LEEPMultiTower.osm [hunk 20]
```diff
@@ -1234,6 +1234,13 @@ OS:Building,
   ,                                       !- Relocatable
   ;                                       !- Nominal Floor to Ceiling Height {m}
 
+OS:AdditionalProperties,
+  {c23ded4b-6f18-4e30-b6db-cc00e1198751}, !- Handle
+  {161a51f0-9700-4aa1-8ba8-f2513d0eecae}, !- Object Name
+  btap_slab_perimeter_m2,                 !- Feature Name 1
+  Double,                                 !- Feature Data Type 1
+  380.23;                                 !- Feature Value 1
```
Comment (brgix) [inline lib/openstudio-standards/standards/necb/NECB2011/data/geometry/LEEPMultiTower.osm:1242]: 3 (problematic) models have AdditionalProperties asserting the slab-on-grade insulation perimeter area:

- LEEPMultiTower
- LEEPTownHouse
- Outpatient

If any 3rd-party model runs into issues (e.g. Boost detecting "inner loops"), this is a straightforward workaround.

Otherwise, this AdditionalProperty is set by `BTAP::Geometry::Spaces.perimeter_m2` (if missing). This is key to establish **costs** and **embodied carbon** linked to slab-on-grade perimeter insulation.

File: Gemfile.lock.3.11.0 [hunk 21]
```diff
@@ -144,7 +144,7 @@ GEM
       yard (~> 0.9, >= 0.9.24)
       yard-activesupport-concern (~> 0.0)
       yard-solargraph (~> 0.1)
-    tbd (3.5.2)
+    tbd (3.6.1)
```
Comment (brgix) [inline Gemfile.lock.3.11.0:147]: Not sure this is the right PR to underline this. _Standards_ should always pull the latest version of TBD.

**Comment by nicholaspneu** [comment]
Merged into NRCan here: https://github.com/NatLabRockies/openstudio-standards/pull/2144
